### ISSUE TITLE
Idiomatic Special Tokenization

### ISSUE THREAD

**Issue by sempervictus** [guoqingbao/vllm.rs#259]
For record-keeping, this is a no-op for @guoqingbao other than reviews as the work progresses.

Model chat templates are evolving to include recursion on tool calls with tool call responses, a variety of media, padding, and specialized functions which we can leverage to improve user experience especially now that we can enforce constraints to force inference iterations in one direction or another.

Secondly, and more importantly, successful generation completes at EOS and there are more and more EOS tokens of various types. When we mask logits we _must_ account for all possible ways in which the model may choose to end its turn or the result is an infinite stream of emojis/run-on repetition of the last few decoded tokens starting from the sequence position at which the model **would** have emitted an EOS.

#232 is not possible without proper handling for that second part so it includes a `SpecialTokens` implementation (and complementary binary tool) which dynamically extracts these from the `Tokenizer` and provides them as idiomatic vectors of ID/string sets for access by the rest of the code. Not only does this help to address the concerns above but rather importantly it also reduces the number of places in the code where we keep strings which break tool calls when they are included making coding agents and the people running them **extremely** unhappy during late-night work efforts...

This issue is here to help track where and how we use special tokens, what's needed to enable the impl to service all of those consumers, and the slow progression of replacing legacy single-token assumptions with vectorized special tokens implementations gracefully having a single item in the vec. This is a long-term effort intended to ease development, support for new models, new idioms/IO paradigms, and replace one-off patchwork with intentional design for idiomatic access.

### Special Token Classes Requiring Redress

- [x] EOS - #232 
- [x] Tool-Call Envelopes (some include binary prefixes requiring very specialized handling in Lark) #260 
- [x] Reasoning (should be able to "induce thinking" via guidance for complex tasks) #262 
- [ ] Media
